File: sdks/local-rp/dart/test/conformance/dns_test.dart
import 'package:linkkeys_local_rp/linkkeys_local_rp.dart';
import 'package:test/test.dart';

import '../testutil/fixtures.dart';

void main() {
  group('dns.json conformance', () {
    test('linkkeys_txt cases', () {
      final d = (loadJson('dns.json') as Map<String, dynamic>)['linkkeys_txt']
          as Map<String, dynamic>;

      for (final c in (d['valid_cases'] as List).cast<Map<String, dynamic>>()) {
        final txt = c['txt'] as String;
        final record = parseLinkKeysTxt(txt);
        final expected = (c['expected_fingerprints'] as List).cast<String>();
        expect(record.fingerprints, equals(expected), reason: 'txt=$txt');
      }

      for (final c
          in (d['invalid_cases'] as List).cast<Map<String, dynamic>>()) {
        final txt = c['txt'] as String;
        try {
          parseLinkKeysTxt(txt);
          fail('expected DnsParseError for txt=$txt');
        } on DnsParseError catch (e) {
          expect(e.symbol, equals(c['expected_error']));
        }
      }

      expect(
          (d['no_record_case'] as Map<String, dynamic>)['documentation_only'],
          isTrue);
    });

    test('linkkeys_apis_txt cases', () {
      final d = (loadJson('dns.json')
          as Map<String, dynamic>)['linkkeys_apis_txt'] as Map<String, dynamic>;

      for (final c in (d['valid_cases'] as List).cast<Map<String, dynamic>>()) {
        final txt = c['txt'] as String;
        final apis = parseLinkKeysApisTxt(txt);
        expect(apis.tcp, equals(c['expected_tcp']), reason: 'txt=$txt');
        expect(apis.httpsBase, equals(c['expected_https_base']),
            reason: 'txt=$txt');
      }

      for (final c
          in (d['invalid_cases'] as List).cast<Map<String, dynamic>>()) {
        final txt = c['txt'] as String;
        try {
          parseLinkKeysApisTxt(txt);
          fail('expected DnsParseError for txt=$txt');
        } on DnsParseError catch (e) {
          expect(e.symbol, equals(c['expected_error']));
        }
      }

      final root = loadJson('dns.json') as Map<String, dynamic>;
      expect(root['default_tcp_port'], equals(defaultTcpPort));
    });
  });
}

File: sdks/local-rp/php/tests/conformance/DnsTest.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/../bootstrap.php';

use LinkKeys\LocalRp\Dns;
use LinkKeys\LocalRp\DnsParseError;

$vectors = loadJson(__DIR__ . '/../../../conformance/dns.json');

foreach ($vectors['linkkeys_txt']['valid_cases'] as $i => $case) {
    TestKit::test("dns.linkkeys_txt.valid.{$i}", function () use ($case) {
        $fps = Dns::parseLinkKeysTxt($case['txt']);
        TestKit::assertEquals($case['expected_fingerprints'], $fps);
    });
}

foreach ($vectors['linkkeys_txt']['invalid_cases'] as $case) {
    TestKit::test('dns.linkkeys_txt.invalid.' . $case['name'], function () use ($case) {
        try {
            Dns::parseLinkKeysTxt($case['txt']);
            throw new \RuntimeException('expected a DnsParseError');
        } catch (DnsParseError $e) {
            TestKit::assertEquals($case['expected_error'], $e->kind);
        }
    });
}

TestKit::test('dns.linkkeys_txt.no_record_case_is_documentation_only', function () use ($vectors) {
    TestKit::assertTrue($vectors['linkkeys_txt']['no_record_case']['documentation_only'] === true);
    TestKit::assertTrue($vectors['linkkeys_txt']['no_record_case']['txt'] === null);
});

foreach ($vectors['linkkeys_apis_txt']['valid_cases'] as $case) {
    TestKit::test('dns.linkkeys_apis_txt.valid.' . $case['name'], function () use ($case) {
        $apis = Dns::parseLinkKeysApisTxt($case['txt']);
        TestKit::assertEquals($case['expected_tcp'], $apis['tcp']);
        TestKit::assertEquals($case['expected_https_base'], $apis['https_base']);
    });
}

foreach ($vectors['linkkeys_apis_txt']['invalid_cases'] as $case) {
    TestKit::test('dns.linkkeys_apis_txt.invalid.' . $case['name'], function () use ($case) {
        try {
            Dns::parseLinkKeysApisTxt($case['txt']);
            throw new \RuntimeException('expected a DnsParseError');
        } catch (DnsParseError $e) {
            TestKit::assertEquals($case['expected_error'], $e->kind);
        }
    });
}

TestKit::test('dns.default_tcp_port', function () use ($vectors) {
    TestKit::assertEquals($vectors['default_tcp_port'], Dns::DEFAULT_TCP_PORT);
});

exit(TestKit::summary('DnsTest') === 0 ? 0 : 1);
